### ISSUE TITLE
serde support

### ISSUE THREAD

**Issue by hicqu** [stepancheg/rust-protobuf#266]
I think `serde_json` is a useful feature. If it's opened, the generated code should be like:
```rust
#[derive(Serialize, Deserialize)]
struct S {
    a: i64,
}
```
So that we can parse `{"a": 100}` to the struct directly.

Current status:

- [x] Basic implementation
- [x] Repeated message fields
- [ ] `Bytes`

**Comment by stepancheg** [comment]
I think two thoughts:

* I'm opposed to adding `#[derive]` unconditionally: currently rust-protobuf has no mandatory dependencies, and I'd like to keep it that way. However, I think it would probably be useful to have codegen option to add those derives (FYI, currently rust-protobuf has an option to use bytes crate).
* I think rust-protobuf should suport JSON encoding and decoding natively. That encoding should follow [official specification](https://developers.google.com/protocol-buffers/docs/proto3#json)

**Comment by bachp** [comment]
I think deriving `Serialize` would be useful to use the generated types in Template generators like [](https://crates.io/crates/handlebars) or [Tera](https://crates.io/crates/tera). They can automatically render types that implement `Serialize`.

Hiding this behind a feature would be fine tough.

**Comment by repi** [comment]
This would indeed be nice as an optional feature!

**Comment by stepancheg** [comment]
Actually now I think that probably it's possible to have it always on.

E. g. generated files could be like this:

```
#[cfg_attr(protobuf_serde, derive(Serialize, Deserialize))]
struct GeneratedMessage {
   ...
}
```

So users who need serde could simply enable flag, and codegen flag won't be needed.

However! I believe JSON serialization and parsing should be done without serde (because it can be done better/more protobuf standard conforming than serde).

I've almost finished reflection in protobuf (now protobuf reflection can mutate messages, not just read), and I plan to implement JSON parsing/serialization in protobuf soon.

**Comment by stepancheg** [comment]
Two additions.

* protobuf now natively supports JSON serialization, serde is probably not that important
* I've created an [issue](https://github.com/stepancheg/rust-protobuf/issues/338) for code generation callbacks, so serde annotations could be inserted in these callbacks

**Comment by stepancheg** [comment]
The current state of serde support:

* `protobuf` crate has `with-serde` feature
* no serde annotations are generated by default
* annotations are generated if `serde_derive` codegen option is specified
* annotations are guarded with `#[cfg_attr(smth, ...)]` if `serde_derive_cfg=smth` option is specified (this part is done mostly for implementation of `with-serde` feature, not sure it is needed for users)

This API design is not final. Feedback is welcome!

**Comment by stepancheg** [comment]
Published version 2.1.0.
